### PR TITLE
libc/ipc: add ftok(3) support

### PR DESCRIPTION
**Pull request by anchao**
## Summary

1. libc/ipc: add ftok(3) support

https://man.openbsd.org/ftok.3

2. sys/ipc: correct definition value of IPC_*

the definition of MSG_NOERROR and IPC_NOWAIT should not be aligned

------------------------------------------------------------------

```
MSGOP(2)

NAME
       msgrcv, msgsnd - System V message queue operations
...
EXAMPLE
       The program below demonstrates the use of msgsnd() and msgrcv().
...
           if (msgrcv(qid, (void *) &msg, sizeof(msg.mtext), msgtype,
                      MSG_NOERROR | IPC_NOWAIT) == -1) {
...

```

### CODE REVIEW

File: include/sys/types.h [hunk 1]
```diff
@@ -158,7 +158,7 @@ typedef int          id_t;
  * semaphores. A key is simply an integer of type key_t
  */
 
-typedef int16_t      key_t;
+typedef long         key_t;
```
[the remark this answers is not in the document]
Comment (anchao) [reply include/sys/types.h]: Done

File: include/sys/ipc.h [hunk 2]
```diff
@@ -35,10 +35,10 @@
 
 /* Mode bits:  The lower order 9-bit bits are the standard mode bits */
 
-#define IPC_MODE    0x01ff     /* Mode bit mask */
-#define IPC_CREAT   (1 << 10)  /* Create entry if key does not exist */
-#define IPC_EXCL    (1 << 11)  /* Fail if key exists */
-#define IPC_NOWAIT  (1 << 12)  /* Error if request must wait */
+#define IPC_MODE   0x0fff /* Mode bit mask */
```
[the remark this answers is not in the document]
Comment (anchao) [reply include/sys/ipc.h:38]: ipc mode is a private implementation of nuttx, which is used as a mask in the source code. dur to changes in other IPC_* definitions, IPC_MODE also needs to be modified accordingly

**Review by pkarashchenko — APPROVED** [review]

File: include/sys/ipc.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (anchao) [reply include/sys/ipc.h:38]: sorry for confuses, MODE mask should be aligned to mode permissions, so the change here is wrong, let me restore this change
https://github.com/apache/incubator-nuttx/blob/master/include/sys/stat.h#L50-L63

File: include/sys/types.h [hunk 3]
```diff
@@ -158,7 +158,7 @@ typedef int          id_t;
  * semaphores. A key is simply an integer of type key_t
  */
 
-typedef int16_t      key_t;
+typedef uint32_t     key_t;
```
Comment (xiaoxiang781216) [inline include/sys/types.h:161]: uint32_t -> int32_t

File: include/sys/types.h [hunk 4]
```diff
@@ -158,7 +158,7 @@ typedef int          id_t;
  * semaphores. A key is simply an integer of type key_t
  */
 
-typedef int16_t      key_t;
+typedef int          key_t;
```
[the remark this answers is not in the document]
Comment (anchao) [reply include/sys/types.h]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**